proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,207 @@
+# 3rd proposal formal spec overview
+
+This is an overview of the 3rd proposal's formal spec additions, to aid in discussions concerning the proposed semantics.
+
+## Abstract Syntax
+
+### Types
+
+#### Tag Types
+
+```
+tagtype ::= [t*]→[]
+```
+
+### Instructions
+
+```
+instr ::= ... | throw tagidx | rethrow labelidx
+        | try blocktype instr* (catch tagidx instr*)* (catch_all instr*)? end
+        | try blocktype instr* delegate labelidx
+```
+
+### Modules
+
+#### Tags
+
+```
+tag ::= export* tag tagtype  | export* tag tagtype import
+```
+
+#### Modules
+
+```
+mod ::= module ... tag*
+```
+
+## Validation (Typing)
+
+#### Modification to labels
+
+To verify that the `rethrow l` instruction refers to a label surrounding the instructions of a catch block (call this a catch-label), we introduce a `kind` attribute to labels in the validation context, which is set to `catch` when the label is a catch-label and empty otherwise.
+
+```
+labelkind ::= catch
+labeltype ::= {result resulttype, kind labelkind?}
+C ::= {..., labels labeltype}
+```
+
+The original notation `labels [t*]` is now an abbreviation for:
+
+```
+labels [t*] ::= labels {result [t*], kind ε}
+```
+
+
+### Instructions
+
+
+```
+C.tags[x] = [t*]→[]
+-----------------------------
+C ⊢ throw x : [t1* t*]→[t2*]
+
+
+C.labels[l].kind = catch
+----------------------------
+C ⊢ rethrow l : [t1*]→[t2*]
+
+
+C.types[bt] = [t1*]→[t2*]
+C, labels [t2*] ⊢ instr* : [t1*]→[t2*]
+(C.tags[x] = [t*]→[] ∧
+ C, labels { result [t2*], kind catch } ⊢ instr'* : [t*]→[t2*])*
+(C, labels { result [t2*], kind catch } ⊢ instr''* : []→[t2*])?
+-----------------------------------------------------------------------------
+C ⊢ try bt instr* (catch x instr'*)* (catch_all instr''*)? end : [t1*]→[t2*]
+
+
+C.types[bt] = [t1*]→[t2*]
+C, labels [t2*] ⊢ instr* : [t1*]→[t2*]
+C.labels[l] = [t*]
+-------------------------------------------
+C ⊢ try bt instr* delegate l : [t1*]→[t2*]
+```
+
+## Execution (Reduction)
+
+### Runtime structure
+
+#### Stores
+
+```
+S ::= {..., tags taginst*}
+```
+
+#### Tag Instances
+
+```
+taginst ::= {type tagtype}
+```
+
+#### Module Instances
+
+```
+m ::= {..., tags tagaddr*}
+```
+
+#### Administrative Instructions
+
+```
+instr ::= ... | throw tagaddr | catch_n{ tagaddr? instr* }* instr* end
+        | delegate{ labelidx } instr* end | caught_m{ tagaddr val^n } instr* end
+```
+
+#### Block contexts and label kinds
+
+So far block contexts are only used in the reduction of `br l` and `return`, and only include labels or values on the stack above the hole. If we want to be able to break jumping over try-catch and try-delegate blocks, we must allow for the new administrative control instructions to appear after labels in block contexts.
+
+```
+B^0 ::= val* [_] instr*
+B^k ::= catch_m{ tagaddr^? instr* }* B^k end | caught_m{ tagaddr val* } B^k end
+      | delegate{ labelidx } B^k end
```

Comment:
By the way does this support `val*` between `label_n` and `catch_m`/`caught_m`/`delegate`? For example,
```
val* (label_n{instr*} (catch_m{ tagaddr? instr* }* B^k end) end
```
This is supported, but
```
val* (label_n{instr*} val* (catch_m{ tagaddr? instr* }* B^k end) end
```
is this supported? Or should we?

Reply:
> This is supported, but
> ```
>  val* (label_n{instr*} val* (catch_m{ tagaddr? instr* }* B^k end) end
> ```
> is this supported? Or should we?

The formal-overview does support this, but I can't see any way to reduce to it.
AFAICT I don't think it is a problem to support this case, even if it were possible.

Reply:
My question was, I didn't think the current block context rules can produce 
```
val* (label_n{instr*} val* (catch_m{ tagaddr? instr* }* B^k end) end
```
meaning `val*` between a `label` and `catch`/`caught`/`delegate`. I might have been mistaken. But I think what was suggested in https://github.com/WebAssembly/exception-handling/pull/143#discussion_r759888458 can create this sequence, so I don't think this will be a problem.